### PR TITLE
feat(plugins): add support for packer snapshots

### LINKED ISSUE #2343: Packer snapshot integration
### Feature Description

Replace/integrate the current workflow for pinning plugins with the new snapshot implementation `:h packer.snapshot()`. 

The snapshot file is in json (even though the documentation states that they're in lua).

### PR DESCRIPTION
**Pull request by kylo252**
# Description

Replace/integrate the current workflow for pinning plugins with the new snapshot implementation, `:h packer.snapshot()`. 

_Note: The snapshot file is in `json` (even though the documentation states that they're in `lua`)._

Fixes #2343
Related #2085

## Features

1. add a default snapshot, `snapshots/default.json`, to hold the the commits of all the core plugins
2. `LvimSyncCorePlugins` now uses `:PackerSnapshotRollback`
3. reduce errors caused by breaking changes in user plugins since they're not handled anymore with `packer.sync()`
4. `PackerSync` will now ignore all the core plugins which are already handled with snapshots, this makes it signficantly faster to finish.
5. allow "unlocking" all the core plugins, by setting an environment variable `$LVIM_DEV_MODE`, e.g. can be defined in `~/.local/bin/lvim`
6. `$LUNARVIM_CACHE_DIR/snapshots/` can be used to store complete snapshots of _all_ the installed plugins, `:h packer.snapshot()`.
7. add a new handler to allow updating `snapshots/default.json` that does not rely on packer or lvim's runtime. It could be completely re-written in any other language.

## How Has This Been Tested?

- Check that `:PackerSync` does not include any core plugins
- Added a new test case that would revert one of the plugins to an earlier commit, and proceed to verify that `:LvimSyncCorePlugins` can restore the previous state

https://github.com/LunarVim/LunarVim/blob/0d761db45777b0d86764bf3a545e47ce5ff9986f/tests/specs/plugins_load_spec.lua#L35

### CODE REVIEW

**Comment by abzcoding** [comment]
are you sure the `make generate_new_lockfile` is actually doing sth?
Screen Shot 2022-03-15 at 4 17 04 PM

**Comment by kylo252** [comment]
> are you sure the `make generate_new_lockfile` is actually doing sth? 

Try altering the values in `snapshots/defaults.json`, maybe something like this:
```json
  "schemastore.nvim": {
    "commit": "DUMMY"
  },
```
Now run the script again.

**Comment by abzcoding** [comment]
@kylo252 you might need to rebase this, since we've added a lot of new changes to the `rolling` branch

**Comment by kylo252** [comment]
@abzcoding, I've updated the PR description, let me know if you think I missed something 😅

**Review by abzcoding — APPROVED** [review]
LGTM 🚀 

we should probably do one last minor release before merging this,

**Comment by kylo252** [comment]
> we should probably do one last minor release before merging this,

@abzcoding, maybe update the screenshot to show alpha.nvim instead as well? 😄

**Comment by abzcoding** [comment]
@kylo252 now we can merge this 🎉

**Comment by kylo252** [comment]
> now we can merge this 🎉

I'm also pretty hopeful it would solve https://github.com/LunarVim/LunarVim/issues/2164

File: lua/lvim/plugin-loader.lua [hunk 1]
```diff
@@ -81,16 +91,14 @@ function plugin_loader.load(configurations)
...
     packer.startup(function(use)
       for _, plugins in ipairs(configurations) do
         for _, plugin in ipairs(plugins) do
           use(plugin)
         end
       end
     end)
-    -- colorscheme must get called after plugins are loaded or it will break new installs.
-    vim.g.colors_name = lvim.colorscheme
-    vim.cmd("colorscheme " .. lvim.colorscheme)
```
Comment (MaxVerevkin) [inline lua/lvim/plugin-loader.lua:93]: This breaks the colorscheme...

Comment (abzcoding) [reply lua/lvim/plugin-loader.lua:93]: https://github.com/LunarVim/LunarVim/pull/2392

**Pull request merged**